fbs: FBS WebRtcTransport

jmillan opened the pull request:
New FBS types:

* CryptoSuite
* SrtpParameters
* FingerprintAlgorithm
* DtlsRole
* IceCandidateType
* IceCandidateTcpType
* IceRole
* IceState

Extra:
- [x] Remove IceState::Close from everywhere.

jmillan committed:
fbs: FBS WebRtcTransport

jmillan commented on node/src/tests/test-WebRtcTransport.ts:
```diff
@@ -306,6 +311,7 @@ test('webRtcTransport.connect() with wrong arguments rejects with TypeError', as
 		fingerprints :
 		[
 			{
+				// @ts-ignore.
 				algorithm : 'sha-256000',
```
Now that we have types, does it make any sense to test with wrong values that the typing system would not allow to happen?

I know some of the values come from the network, but the client side is also typed, so IMHO we should consider undefined behaviour if wrong values are provided in runtime. Otherwise our code needs to consider such values and not simply rely on the types, which defeats the purpose of the type system IMHO.

ibc replied:
> Now that we have types, does it make any sense to test with wrong values that the typing system would not allow to happen?

Yes, because we cannot assume that every user uses TypeScript or validates network received data no matter it's created by media soup-client or not.

jmillan replied:
I think we should at some point. Ie: in Rust we don't do it that way and we rely on the type system.

ibc replied:
In Rust you cannot pass an invalid enum or type, in JS you can.

jmillan replied:
> In Rust you cannot pass an invalid enum or type, in JS you can.

>> or validates network received data no matter it's created by media soup-client or not.

Such argument applies to rust too.

ibc replied:
In Rust you **cannot** pass a random string.

jmillan replied:
> In Rust you cannot pass a random string.

I'm talking about runtime. We can't control what's coming from the network.

ibc replied:
I don't care what we receive from the network in a Rust app. Whatever we receive from the network we have to convert into a proper `webrtc_transport_options` Rust type **first** before we pass it to `router.create_webrtc_transport()`.

jmillan replied:
This is not being done. I'm removing the `skip` from the test and properly throw when receiving wrong data.

jmillan commented on worker/include/RTC/DtlsTransport.hpp:
```diff
@@ -28,7 +30,6 @@ namespace RTC
 	public:
 		enum class Role
 		{
-			NONE = 0,
```
Removed in favour of `std::optional`

jmillan replied:
Do not use `NONE` as an internal way to indicate that there is no `Role`. Make use of c++17 `std::optional` functionality.

ibc replied:
Good.

jmillan commented on worker/include/RTC/DtlsTransport.hpp:
```diff
@@ -37,7 +38,6 @@ namespace RTC
 	public:
 		enum class FingerprintAlgorithm
 		{
-			NONE = 0,
```
Removed in favour of `std::optional`

jmillan commented on worker/include/RTC/IceServer.hpp:
```diff
@@ -17,9 +18,14 @@ namespace RTC
 			COMPLETED,
-			DISCONNECTED
+			DISCONNECTED,
+			CLOSED
```
Added `CLOSED` for consistency with Node and Rust. This can probably be removed though.

I'll review this.

ibc replied:
When is CLOSED needed?

jmillan replied:
It is not. But it's present in TS and RUST. I'll create a TODO to remove them in this PR.

jmillan replied:
Done.

jmillan commented on worker/src/RTC/DtlsTransport.cpp:
```diff
@@ -149,6 +149,101 @@ namespace RTC
 			SSL_CTX_free(DtlsTransport::sslCtx);
 	}
 
+	DtlsTransport::Role DtlsTransport::RoleFromFbs(FBS::WebRtcTransport::DtlsRole role)
```
Please @ibc, can you take a look at this file?

I've reviewed multiple times and all tests pass.

ibc replied:
It's a simple 1:1 mapping between C++ enums and FBS enums, what can be wrong? :)

jmillan replied:
I meant `file` not block :-)

ibc replied:
ok, next time click here to add a file comment:

![Screenshot 2023-09-11 at 12 19 37](https://github.com/versatica/mediasoup/assets/16191/b64fc7b1-8630-4d60-96f0-99d38f8a4c3b)

ibc replied:
File looks good.

ibc commented on worker/src/Channel/ChannelSocket.cpp:
```diff
@@ -259,9 +259,9 @@ namespace Channel
 		// TMP: For debugging.
-		// auto s = flatbuffers::FlatBufferToString(
-		//   reinterpret_cast<uint8_t*>(msg), FBS::Message::MessageTypeTable());
-		// MS_ERROR("%s", s.c_str());
+		auto s = flatbuffers::FlatBufferToString(
+		  reinterpret_cast<uint8_t*>(msg), FBS::Message::MessageTypeTable());
+		MS_ERROR("%s", s.c_str());
```
Can we remove this?

jmillan replied:
I'll put it under a flag. It's intended for development and it's extremely useful.

ibc replied:
Please don't use `MS_ERROR`. Use `MS_DUMP`.

ibc commented on worker/src/RTC/DtlsTransport.cpp:
```diff
@@ -913,7 +1011,7 @@ namespace RTC
 			this->timer->Stop();
 
 			// Process the handshake just once (ignore if DTLS renegotiation).
-			if (!wasHandshakeDone && this->remoteFingerprint.algorithm != FingerprintAlgorithm::NONE)
```
Is this condition wrapped into some `if (this->remoteFingerprint.algorithm.has_value())`? It's hard to see in the diff.

jmillan replied:
Nop, `FingerPrint::algorithm` is not optional. If the handshake was done it must have a valid value.

ibc replied:
ok

ibc commented on worker/src/RTC/DtlsTransport.cpp:
```diff
@@ -1031,8 +1129,6 @@ namespace RTC
 		MS_TRACE();
 
 		MS_ASSERT(this->handshakeDone, "handshake not done yet");
-		MS_ASSERT(
```
Here we should assert that `this->remoteFingerprint.algorithm.has_value()` right?

jmillan replied:
Fingerprint.algorithm is not optional.

ibc commented on worker/src/RTC/IceCandidate.cpp:
```diff
@@ -1,39 +1,66 @@
+	IceCandidate::CandidateType IceCandidate::CandidateTypeFromFbs(
+	  FBS::WebRtcTransport::IceCandidateType type)
 	{
-		MS_TRACE();
```
Please don't remove `MS_TRACE()` calls.

jmillan replied:
I didn't. `IceCandidate::FillBuffer` has it's `MS_TRACE`. The diff just shows it moved.

jmillan commented on rust/tests/integration/webrtc_transport.rs:
```diff
@@ -161,6 +160,8 @@ fn create_succeeds() {
                         .try_into()
                         .unwrap(),
                     );
+                    // TODO: I wonder how this was not needed before...
```
Nazar? `enable_sctp` must be true and it was not explicit before. I wonder how was it working before. I don't see what I'm missing.

nazar-pc replied:
Not sure, was it set in TypeScript before?

jmillan commented on rust/tests/integration/webrtc_transport.rs:
```diff
@@ -194,6 +195,29 @@ fn create_succeeds() {
+                assert!(ice_candidates[0].priority > ice_candidates[1].priority);
+                assert!(ice_candidates[1].priority > ice_candidates[2].priority);
+
+                // TODO: How comes UDP and TCP where created before?
```
We were creating explicitly UDP sockets, how could we be expecting UDP and TCP? cc @nazar-pc.

nazar-pc replied:
IIRC `ListenInfo` is a relatively new thing, apparerntly something was off there? Check git history for this section of the code, I do not remember off top of my head.

ibc replied:
What do we mean here? Is the test expecting a TCP socket even if only UDP was requested? Is this a bug in Rust create_webrtc_transport? In Rust webrtc_transport constructor? Or just in the test?

nazar-pc replied:
If test was passing before worker must have been creating TCP socket in the past as well. I don't think Rust side was hallucinating things that were not there. Maybe it was fixed with flatbuffers update? IDK.

ibc replied:
I migrated `ListenIps` to `ListenInfos` in Rust entirely (while in Node we still allow `listenIps` for backwards compatibility) so it's imposible that Rust side is creating TCP sockets if not requested by the app.

ibc replied:
Wow, something must be wrong since the test is indeed just creating UDP sockets:

```rust
let transport1 = router
                .create_webrtc_transport({
                    let mut webrtc_transport_options = WebRtcTransportOptions::new(
                        vec![
                            ListenInfo {
                                protocol: Protocol::Udp,
                                ip: IpAddr::V4(Ipv4Addr::LOCALHOST),
                                announced_ip: Some("9.9.9.1".parse().unwrap()),
                                port: None,
                                send_buffer_size: None,
                                recv_buffer_size: None,
                            },
                            ListenInfo {
                                protocol: Protocol::Udp,
                                ip: IpAddr::V4(Ipv4Addr::UNSPECIFIED),
                                announced_ip: Some("9.9.9.2".parse().unwrap()),
                                port: None,
                                send_buffer_size: None,
                                recv_buffer_size: None,
                            },
                            ListenInfo {
                                protocol: Protocol::Udp,
                                ip: IpAddr::V4(Ipv4Addr::LOCALHOST),
                                announced_ip: None,
                                port: None,
                                send_buffer_size: None,
                                recv_buffer_size: None,
                            },
                        ]
                        .try_into()
                        .unwrap(),
                    );
                    webrtc_transport_options.num_sctp_streams = NumSctpStreams {
                        os: 2048,
                        mis: 2048,
                    };
                    webrtc_transport_options.max_sctp_message_size = 1000000;
                    webrtc_transport_options.app_data = AppData::new(CustomAppData { foo: "bar" });

                    webrtc_transport_options
                })
```

ibc replied:
Sorry, now I don't understand. What is wrong @jmillan? I don't see that the test is expecting TCP sockets. Where do you see that?

nazar-pc replied:
Check commented-out code below

ibc replied:
Oh hell, sorry.

ibc replied:
So the test passes in `v3` and `flatbuffers` branch because:

```
#[test]
#[ignore]
```

¯\_(ツ)_/¯

@jmillan I'm trying to fix it in this branch directly.

ibc replied:
Guys, there is nothing to test. The test was disabled in `v3` and `flatbuffers` branches, and in this branch it is enabled, fixed and it passes. That's all.

@jmillan I'll remove the commented-out code and push after confirming everything is ok.

ibc replied:
Pushed. Resolving this thread.

jmillan committed:
Node: lint

jmillan committed:
Merge remote-tracking branch 'origin/flatbuffers' into webrtc-transport-flatbuffers

jmillan committed:
fbs.rs: add missing innier line doc

It was generating a clippy error

jmillan commented:
**EDIT**: After merging `flatbuffers` the wrong line appeared and it's now fixed by a new commit.

worker linting is failing, giving this error:

<img width="1139" alt="Screenshot 2023-09-12 at 12 00 28" src="https://github.com/versatica/mediasoup/assets/732340/8a1a6c70-e096-43f8-8703-28b92341dee1">

Such line in this branch does NOT look like the one in the red block! Do you have any idea of what can be happening here?

jmillan committed:
lint

nazar-pc commented on rust/src/data_structures.rs:
```diff
@@ -389,6 +477,93 @@ pub enum DtlsFingerprint {
+            web_rtc_transport::FingerprintAlgorithm::Sha1 => {
+                let mut value_result = [0_u8; 20];
+
+                hex_as_bytes(fingerprint.value.as_str(), &mut value_result);
```
Why is it sent as hex and not binary from worker in the first place?

jmillan replied:
This is probably because the client needs to use it in hex mode, so everyone in the chain, starting from worker, provides it in this form. I wouldn't be totally against having it in binary form in server side in a future.

ibc commented on node/src/WebRtcTransport.ts:
```diff
@@ -744,8 +848,10 @@ function parseIceCandidate(binary: FbsWebRtcTransport.IceCandidate): IceCandidat
-		type       : 'host',
-		tcpType    : binary.tcpType() === 'passive' ? 'passive' : undefined
+		type       : iceCandidateTypeFromFbs(binary.type()),
+		tcpType    : binary.tcpType() === null ?
```
Here we are relying on `null` value. However we want it to be `undefined` when not present in the FBS object. If in the future this feature is accepted (you wrote a PR AFAIR) this is gonna fail. So wouldn't it be better if ew check `!binary.tcpType()` instead?

jmillan replied:
TS flatbuffers does provide `null` rather than `undefined` for optionals.

> So wouldn't it be better if ew check !binary.tcpType() instead?

Nop, because the corresponding enum value can be 0, and thus such check cannot be done as you would get a false negative.

ibc replied:
I know that TS flatbuffers generates null instead of undefined, but we consider that unfortunate and you wrote a PR changing that, am I wrong?

ibc replied:
> Nop, because the corresponding enum value can be 0, and thus such check cannot be done as you would get a false negative.

Right. What about checking "typeof xxxx === "number"?

Yes, I insist in being ready for the lib to produce undefined instead of null.

jmillan replied:
> Right. What about checking "typeof xxxx === "number"?

I have nothing agains that. We can do this in a separate PR for all optional values.

I made a [feature request](https://github.com/google/flatbuffers/issues/7656), but I'm not planning into doing the PR myself mid term.

ibc replied:
Ah ok, I thought you wrote a PR. We are good then.

nazar-pc commented on rust/src/data_structures.rs:
```diff
@@ -389,6 +477,93 @@ pub enum DtlsFingerprint {
 }
 
+fn hex_as_bytes(input: &str, output: &mut [u8]) {
+    for (i, v) in output.iter_mut().enumerate() {
+        *v = u8::from_str_radix(&input[i * 3..(i * 3) + 2], 16)
+            .map_err(|error| {
+                panic!("Failed to parse value {input} as series of hex bytes: {error}")
+            })
+            .unwrap();
+    }
```
Generally try to avoid manual indexing in loops in Rust, use higher level operations to get an iterator and zip it with the other iterator to move in lock step. This also typically results in vectorized machine code that is good for performance too.

Something like this:
```suggestion
    for (i, o) in input.splot(':').zip(&mut output) {
        *o = u8::from_str_radix(i, 16)
            .unwrap_or_else(|error| {
                panic!("Failed to parse value {i} as series of hex bytes: {error}")
            });
    }
```

jmillan replied:
applied, thanks.

nazar-pc commented on rust/src/data_structures.rs:
```diff
@@ -826,6 +1001,30 @@ pub struct DtlsParameters {
+impl DtlsParameters {
+    pub(crate) fn to_fbs(&self) -> web_rtc_transport::DtlsParameters {
+        web_rtc_transport::DtlsParameters {
+            role: Some(self.role.to_fbs()),
```
Why is `role` optional in flatbuffers?

ibc replied:
AFAIR we allow it to be optional so we decide in worker side.

nazar-pc replied:
Not sure what benefit it has because `None` case now has to be handled by both worker and library (we have `unwrap_or` below). It would be easier for everyone if it wasn't optional in the first place.

ibc replied:
I may be wrong but role None is used to indicate that this->role in DtlsTransport.cpp is not yet set (None is its default initial value). However role is optional in arguments given to webRtcTransport.connect() method. Those are different things.

jmillan replied:
IMHO it makes sense that the **DtlsParameters** type has non optional DtlsRole, and having it defaulted to 'auto'.

nazar-pc commented on rust/src/messages.rs:
```diff
@@ -863,6 +885,22 @@ impl RouterCreateWebrtcTransportRequest {
+    pub(crate) fn to_fbs(&self) -> web_rtc_transport::WebRtcTransportOptions {
+        web_rtc_transport::WebRtcTransportOptions {
+            base: Box::new(transport::Options {
+                direct: false,
+                max_message_size: 0,
```
Why do we have these fields in "base" options that do not make sense to generic transport and set to fake values?

jmillan replied:
I'm setting this to optional. Both 'direct'  and 'max_message_size'  are used by the base Transport class and only applied to direct transport.

I'm setting it to optional in this PR.

ibc replied:
AFAIR we inspect and use those fields in Transport.cpp so they need to be in the base options of message.

jmillan replied:
Done.

nazar-pc commented on rust/src/router/webrtc_transport.rs:
```diff
@@ -851,10 +973,10 @@ impl WebRtcTransport {
         let response = self
             .inner
             .channel
-            .request(self.id(), TransportRestartIceRequest {})
+            .request_fbs(self.id(), TransportRestartIceRequest {})
             .await?;
 
-        Ok(response.ice_parameters)
+        Ok(response)
```
```suggestion
        self
            .inner
            .channel
            .request_fbs(self.id(), TransportRestartIceRequest {})
            .await
```

nazar-pc commented on rust/src/router/webrtc_transport.rs:
```diff
@@ -201,26 +286,27 @@ pub struct WebRtcTransportStat {
     pub available_incoming_bitrate: Option<u32>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub max_incoming_bitrate: Option<u32>,
+    pub max_incoming_bitrate: u32,
```
Why not an option anymore? I thought it was an option that may not be set, in which case it would be `None`.

jmillan replied:
It should be an option and be using `std::optional` in C++ in the first place. But it's initialized to 0 instead. We need to review all the places in the worker where `std::optional` make sense.

I'll make this field optional in this PR.

jmillan replied:
done.

nazar-pc reviewed (commented):
Only looked at Rust changes

ibc committed:
Rust webrtc_transport.rs: remove old code

jmillan committed:
enhance hex_as_bytes function

jmillan committed:
simplify restart_ice() method

jmillan committed:
Enhance transport options

Make max_message_size and initial_available_outgoing_bitrate optional.

jmillan committed:
Enhance transport stats

Make needed fields optional.

ibc committed:
DtlsTransport::Run(): Fix wrong std::optional access

nazar-pc commented:
In order to minimize notifications, please just resolve comments, no need to comment "done" on each one. I get way too many notifications of various kind from mediasoup repo. Thanks for understanding :pray:

ibc committed:
Fix node/src/utils.ts clone() usage

Since a recent commit, it no longer returns an empty object is `undefined` is given but instead it returns `undefined`.

jmillan committed:
Add debugging MS_DUMP under flag

ibc committed:
Remove unnecesary has_value() check

jmillan committed:
C++: Remove unneeded 'default' switch statements

All options are exhausted already.

This is because we don't use NONE enums anymore.

jmillan committed:
Merge remote-tracking branch 'origin/flatbuffers' into webrtc-transport-flatbuffers

jmillan committed:
FBS: Make DtlsParameters.role mandatory

ibc commented on worker/include/RTC/DtlsTransport.hpp:
```diff
@@ -2,12 +2,14 @@
 #include <openssl/ssl.h>
 #include <openssl/x509.h>
 #include <absl/container/flat_hash_map.h>
+#include <optional>
```
Please don't include this. It's already included in `common.hpp`.

ibc replied:
I'll remove it.

ibc replied:
No, I can't due to (explained in Slack).

jmillan committed:
C++: remove include statement

optional already included in common.hpp
